imr: make the transition to overloaded size_of context functions

eliransin opened the pull request:
This mini set, makes the transition to overloaded size_of functions in context instead of templated ones.
This patchset does this in 2 stages:
1. First the transition is made trivially - we keep the templated implementations but changing the constrainet to an overloaded one. This means that the template implementation in the contexts generates the overloaded functions.
2. We drop the templated implementations and only implement the size_of overloaded versions that are actually needed. This adds the safety that was intended in the first place, whenever a size_of for a specific tag is not implemented the compilation will fail with constraint not met compilation error instead of just generating a default version using the template.

Commit:
imr fundumentals: Create a concept for having a size_of<Tag> function

This commit defines the requirement of having a size_of<Tag> function
in a HasSizeof concept and then uses it whenever this is required.
This is done as a preparation for changing the requirement to
have an overloaded size_of (size_of(Tag)) instead of a templated
function.

Commit:
imr fundumentals: Move to overloaded size_of

We have done the following in the context of the move:
1. The requirement for having a size_of<Tag>() function was
replaced with having a size_of(Tag). i
2. The imr templates that call size_of<Tag>() have been changed to call
   size_of(Tag{})
3. Tags of imr object that act as the Tag part of the requirement
have been made complete types by adding {} to the end of their
declaration - making it a concrete definition.
4. The Context objects that are now failing the constrains were
   trivially made compatible with the constraint by adding the `Tag`
   typename into the empty parenthesis.

All of those stages were done at once so compilation keeps passing.

eliransin commented:
@denesb I have no permissions to ask for code review on your repo, I don't expect you to change this but only to review this miniset :slightly_smiling_face:

denesb commented on data/cell.hh:
```diff
@@ -53,35 +53,33 @@ struct cell {
+        class chunk_back_pointer {};
+        class chunk_next {};
+        class chunk_data {};
+        class last_chunk_size {};
```
This could have been done as a separate step but this is a hackaton so I won't bikeshed. :grin:

eliransin replied:
You are right, it could but thanks for not making me to separate to another commit :slightly_smiling_face:

denesb commented on data/cell.hh:
```diff
@@ -261,7 +259,7 @@ struct cell {
             }
             template<typename Tag>
-            size_t size_of() const noexcept {
+            size_t size_of(Tag) const noexcept {
```
We should de-template these methods. I think you can just remove the template and it will keep working.

eliransin replied:
As I wrote, this is a trivial transfromation.
But, if I remove the template I will need a size_of(Tag) for every tag that tags a buffer.
Meaning I will need to have separate implementations for:
```
size_t size_of(T1) {
 return ...;
}

size_t size_of(T2) {
return ...;
}
...
```
This is what you wanted in the first place didn't you?

BTW, in the current implementation there is exactly one specialization for this method, meaning that even if I did this separation I will end up with a bunch of identical overloaded functions and one more which is different.
But so I am sure that I am not talking complete nonsense I will remove the template and see if it compile... :smile:

eliransin replied:
@denesb  It didn't work, I got a lot of errors of this style:
```
...
error: no matching function for call to 'imr::buffer<data::cell::tags::data>::serialized_object_size(const uint8_t*&, data::cell::variable_value::context)'
  245 |             return alternative::type::serialized_object_size(in, context.template context_for<typename alternative::tag>(in));
      |                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ./data/cell.hh:30,
                 from ./types.hh:27,
                 from ./dht/i_partitioner.hh:45,
                 from ./dht/sharder.hh:24,
                 from sstables/sstables.cc:39:
./imr/fundamental.hh:260:19: note: candidate: 'static size_t imr::buffer<Tag>::serialized_object_size(const uint8_t*, const Context&) [with Context = data::cell::variable_value::context; Tag = data::cell::tags::data; size_t = long unsigned int; uint8_t = unsigned char]'
  260 |     static size_t serialized_object_size(const uint8_t*, const Context& context) noexcept {
```
Basically it boils down to that:
The compiler searched for a function `size_t size_of(data::cell::tags::data)` in `data::cell::variable_value::context` and couldn't find it. (which in turn didn't generate the `serialized_object_size` function which calls size_of).
We will get this error at least for every concrete tag that is used in a buffer for the corresponding context.

With all of that said, it is your project and if you think this is the way it should be I will do it.
I can also do it in a follow up step and then we will consider if to get this follow up in or not.

eliransin replied:
@denesb ping, what do you decide? Should I do it in this step or should I do it in a follow-up?

denesb replied:
The point of making these functions overloads is to remove the template, because leaving it as a template leaves the possibility of mistakes open. A new buffer can be added, or one changed and it will happily keep using the out-of-date `size_of()` in the context.

> But, if I remove the template I will need a size_of(Tag) for every tag that tags a buffer.

Yes, that would be the point. But don't we already have one of these per buffer already? Just as a template specialization, not an overload?

> Meaning I will need to have separate implementations for:
>```
> size_t size_of(T1) {
> return ...;
> }
>
>size_t size_of(T2) {
> return ...;
> }
> ...
>```
>
> This is what you wanted in the first place didn't you?

Yes. I'm fine with this being done in a follow-up as long as it's going to happen.

Commit:
imr: drop templated size_of completely and move to overloaded form

Now that the imr infrastructure expects the overloaded form of
size_of instead of the templated form, we can drop the templated
form completely. This will give us the extra protection of not
defaulting to some generic template form when the specific overloaded
form of size_of is not implemented.
After this transition, if a matching size_of overload is not
implemented when needed, the compiler will emit constraint not met
error.

eliransin commented:
@avikivity I have added you since botond is not working today and it seamed not right to just go ahead and merge my own PR :slightly_smiling_face:

eliransin commented:
Since @denesb and me already agreed it is ok, I even added what we discussed to be a follow up, I am going ahead and merge this. (fyi @denesb )
The reason is that my next PR will probably be dependent on this.
